precert/corr.go

```diff
@@ -0,0 +1,202 @@
+// this function rejects such pairs.
+func Correspond(precertDER, finalDER []byte) error {
+	preTBS, err := tbsDERFromCertDER(precertDER)
+	if err != nil {
+		return fmt.Errorf("parsing precert: %w", err)
+	}
+
+	finalTBS, err := tbsDERFromCertDER(finalDER)
+	if err != nil {
+		return fmt.Errorf("parsing final cert: %w", err)
+	}
+
+	// The first 7 fields of TBSCertificate must be byte-for-byte identical.
+	// The next 2 fields (issuerUniqueID and subjectUniqueID) are forbidden
+	// by the Baseline Requirements so we assume they are not present (if they
+	// are, they will fail the next check, for extensions).
+	// https://datatracker.ietf.org/doc/html/rfc5280#page-117
+	// TBSCertificate  ::=  SEQUENCE  {
+	//      version         [0]  Version DEFAULT v1,
+	//      serialNumber         CertificateSerialNumber,
+	//      signature            AlgorithmIdentifier,
+	//      issuer               Name,
+	//      validity             Validity,
+	//      subject              Name,
+	//      subjectPublicKeyInfo SubjectPublicKeyInfo,
+	//      issuerUniqueID  [1]  IMPLICIT UniqueIdentifier OPTIONAL,
+	//      					 -- If present, version MUST be v2 or v3
+	//      subjectUniqueID [2]  IMPLICIT UniqueIdentifier OPTIONAL,
+	//      					 -- If present, version MUST be v2 or v3
+	//      extensions      [3]  Extensions OPTIONAL
+	//      					 -- If present, version MUST be v3 --  }
+	for i := 0; i < 7; i++ {
+		if err := readIdenticalElement(&preTBS, &finalTBS); err != nil {
+			return fmt.Errorf("checking for identical field %d: %w", i, err)
+		}
+	}
+
+	// The extensions should be mostly the same, with these exceptions:
+	//  - The precertificate should have exactly one precertificate poison extension
+	//    not present in the final certificate.
+	//  - The final certificate should have exactly one SCTList extension not present
+	//    in the precertificate.
+	//  - As a consequence, the byte lengths of the extensions fields will not be the
+	//    same, so we ignore the lengths (so long as they parse)
+	precertExtensionBytes, err := unwrapExtensions(preTBS)
+	if err != nil {
+		return fmt.Errorf("parsing precert extensions: %w", err)
+	}
+
+	finalCertExtensionBytes, err := unwrapExtensions(finalTBS)
+	if err != nil {
+		return fmt.Errorf("parsing final cert extensions: %w", err)
+	}
+
+	precertParser := extensionParser{bytes: precertExtensionBytes, skippableOID: poisonOID}
+	finalCertParser := extensionParser{bytes: finalCertExtensionBytes, skippableOID: sctListOID}
+
+	for i := 0; ; i++ {
+		precertExtn, err := precertParser.Next()
+		if err != nil {
+			return err
+		}
+
+		finalCertExtn, err := finalCertParser.Next()
+		if err != nil {
+			return err
+		}
+
+		if !bytes.Equal(precertExtn, finalCertExtn) {
+			return fmt.Errorf("extensions differed at position %d: '%x' (precert) vs '%x' (final)",
+				i+precertParser.skipped, precertExtn, finalCertExtn)
```

Comment:
Someone who actually runs into this error might get confused because the error message includes the actual extension index within the precert, but (potentially) lies about the extension index within the final cert. It would be good to incorporate `I=finalCertParser.skipped` as well, but I don't have any good ideas for how to do so without becoming even more confusing :(